### PR TITLE
TraefikTomlProxy

### PR DESCRIPTION
**Pull request by GeorgianaElena**
closes https://github.com/jupyterhub/traefik-proxy/issues/21

The data passed in when adding a route is only stored in memory. I couldn't add it in the toml file because traefik doesn't know how to parse it. @minrk , do you think we should we also store it in a file (other than the two tomls)? 

I also tried to move some of the common api to the base class TraefikProxy and made `test_proxy.py` work for both proxy implementations.

### CODE REVIEW

**Comment by minrk** [comment]
> I couldn't add it in the toml file because traefik doesn't know how to parse it

What happens if you put the data in the toml file as part of the route? Does traefik raise an error? If so, can you put it in another part of the toml file that traefik doesn't look at, perhaps? An additional top-level key?

**Comment by minrk** [comment]
> I also tried to move some of the common api to the base class TraefikProxy and made test_proxy.py work for both proxy implementations.

A+, excellent!

File: .travis.yml [hunk 1]
```diff
@@ -11,11 +11,11 @@ env:
 
 # installing dependencies
 before_install:
+  - pip install --upgrade pytest
```
Comment (minrk) [inline .travis.yml]: Rather than upgrading pytest here, add `--upgrade` to the dev-requirements below.

I also see that dev-requirements is added twice. It doesn't need to be in the `install` step as well as the `before_install` step

**Comment by GeorgianaElena** [comment]
@minrk, quick question. What's the type of `data` we pass to `add_route`? In the method documentation it says it's `dict`, but in jupyterhub `test_proxy` the dict seems to be serialized to str before passing it to `add_route`.

**Comment by GeorgianaElena** [comment]
I managed to put the data into the toml file, yay! I was putting it between double quotes instead of using single quotes (data already contained double quotes).

File: jupyterhub_traefik_proxy/proxy.py [hunk 2]
```diff
@@ -19,11 +19,129 @@
...
+            )
+        elif config_type == "etcd":
+            self.traefik_process = Popen(
+                ["traefik", "--etcd", "--etcd.useapiv3=true"], stdout=None
+            )
+        else:
+            self.log.error(
+                "Configuration mode not supported \n. The proxy can only be configured through toml and etcd"
+            )
+            raise
```
Comment (minrk) [inline jupyterhub_traefik_proxy/proxy.py]: `raise` only works if there's an Exception. Try: `raise ValueError(...)` with the above message.

File: jupyterhub_traefik_proxy/toml.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+
+class TraefikTomlProxy(TraefikProxy):
+    """JupyterHub Proxy implementation using traefik and toml config file"""
+
+    mutex = Any()
+
+    @default("mutex")
+    def _default_mutex(self):
+        return asyncio.Lock()
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: The atomic writing should probably follow the pattern [used in the notebook](https://github.com/jupyter/notebook/blob/5.7.0/notebook/services/contents/fileio.py#L71) where we write a temporary file then move it into place. This way there's never a partial configuration file on disk, which traefik might read while we aren't done writing.

It can be a little simpler than the one in the notebook since we know:

1. that it's not Windows
2. that it's a text file
3. we require Python 3
4. we don't need to turn errors into log messages (raising is ok)

File: jupyterhub_traefik_proxy/toml.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        super().__init__(**kwargs)
+        self.routes_cache = {}
+
+    async def _setup_traefik_static_config(self):
+        self.log.info("Setting up traefik's static config...")
+        self._generate_htpassword()
+
+        try:
+            with open(self.toml_static_config_file, "w") as f:
+                f.write('defaultEntryPoints = ["http"]\n')
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: Instead of writing toml-as-text, let's use the [toml package](https://github.com/uiri/toml) to ensure we always create valid toml.

File: jupyterhub_traefik_proxy/toml.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+                f.write('\tentrypoint = "auth_api"\n'),
+
+                f.write("[file]\n")
+                f.write('\tfilename = "' + self.toml_dynamic_config_file + '"\n')
+                f.write("\twatch = true\n")
+            open(self.toml_dynamic_config_file, "a").close()
+        except IOError as e:
+            self.log.error(
+                "Couldn't set up traefik's static config. I/O error({0}): {1}".format(
+                    e.errno, e.strerrornse
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: If we use `self.log.exception("msg...")`, the logger will take care of formatting the exception itself, including a traceback.

File: jupyterhub_traefik_proxy/toml.py (hunk 3 above)
Comment (minrk) [reply jupyterhub_traefik_proxy/toml.py]: That doesn't mean we don't want this mutex, I realize. We still want that to protect access to the routes with this.

**Comment by GeorgianaElena** [comment]
@minrk, I addressed your comments and done some more refactoring. Also, I shared the etcd fixture to the entire test session because sometimes etcd didn't have enough time to start between the test modules.

File: jupyterhub_traefik_proxy/etcd.py [hunk 6]
```diff
@@ -65,102 +59,61 @@ def _default_client(self):
...
-
-        ht = HtpasswdFile()
-        ht.set_password(self.traefik_api_username, self.traefik_api_password)
-        self.traefik_api_hashed_password = str(ht.to_string()).split(":")[1][:-3]
-
     async def _setup_traefik_static_config(self):
-        self.log.info("Setting up traefik's static config...")
-        self._create_htpassword()
+        await super()._setup_traefik_static_config()
+        keys = []
```
Comment (minrk) [inline jupyterhub_traefik_proxy/etcd.py]: Maybe use a dict instead of lists of keys and values?

File: jupyterhub_traefik_proxy/etcd.py [hunk 7]
```diff
@@ -65,102 +59,61 @@ def _default_client(self):
...
+            KV.put.txn(
+                self.etcd_traefik_prefix + "etcd/prefix", self.etcd_traefik_prefix
+            ),
+            KV.put.txn(self.etcd_traefik_prefix + "etcd/useapiv3", "true"),
+            KV.put.txn(self.etcd_traefik_prefix + "etcd/watch", "true"),
+            KV.put.txn(self.etcd_traefik_prefix + "providersThrottleDuration", "1"),
+        ]
+
+        for key, value in zip(keys, values):
+            if type(value) is bool:
```
Comment (minrk) [inline jupyterhub_traefik_proxy/etcd.py]:
```suggestion
            if isinstance(value, bool):
```

File: jupyterhub_traefik_proxy/etcd.py [hunk 8]
```diff
@@ -65,102 +59,61 @@ def _default_client(self):
...
+                value = str(value).lower()
+            if type(value) is list:
```
Comment (minrk) [inline jupyterhub_traefik_proxy/etcd.py]:
```suggestion
            if isinstance(value, list):
```

File: jupyterhub_traefik_proxy/etcd.py [hunk 9]
```diff
@@ -65,102 +59,61 @@ def _default_client(self):
...
+            KV.put.txn(self.etcd_traefik_prefix + "etcd/endpoint", "127.0.0.1:2379"),
+            KV.put.txn(
+                self.etcd_traefik_prefix + "etcd/prefix", self.etcd_traefik_prefix
+            ),
+            KV.put.txn(self.etcd_traefik_prefix + "etcd/useapiv3", "true"),
+            KV.put.txn(self.etcd_traefik_prefix + "etcd/watch", "true"),
+            KV.put.txn(self.etcd_traefik_prefix + "providersThrottleDuration", "1"),
+        ]
+
+        for key, value in zip(keys, values):
```
Comment (minrk) [inline jupyterhub_traefik_proxy/etcd.py]: If we use a dict, this would be `for key, value in d.items():`

File: jupyterhub_traefik_proxy/etcd.py [hunk 10]
```diff
@@ -280,18 +221,14 @@ def _stop_traefik(self):
...
-            self, frontend_alias
-        )
+        (
+            backend_alias,
+            backend_url_path,
+            backend_weight_path,
+            frontend_alias,
+            frontend_backend_path,
+            frontend_rule_path,
+        ) = traefik_utils.generate_route_keys(self, target, routespec)
```
Comment (minrk) [inline jupyterhub_traefik_proxy/etcd.py]: Maybe for later: returning long tuples can get unwieldy and hard to follow. At this point, it might be worth using a [namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple) to return a single object with named attribute access.

File: jupyterhub_traefik_proxy/toml.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        except FileNotFoundError as e:
+            self.log.error(
+                "Failed to find traefik \n"
+                "The proxy can be downloaded from https://github.com/containous/traefik/releases/download."
+            )
+            raise
+
+    def _clean_resources(self):
+        try:
+            os.remove(self.toml_static_config_file)
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: We'll need to be careful here: we should only delete the 'static' config file if we are writing it (i.e. `if self.should_start`)

File: jupyterhub_traefik_proxy/toml.py [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+            }
+        except KeyError:
+            self.log.info("No route for {} doesn't exist!".format(routespec))
+            result = None
+        finally:
+            return result
+
+    def _persist_routes(self):
+        with traefik_utils.atomic_writing(self.toml_dynamic_config_file) as config_fd:
+            config_fd.write("[frontends]\n")
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: use `toml.dump` here?

File: jupyterhub_traefik_proxy/toml.py [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+    async def _setup_traefik_static_config(self):
+        await super()._setup_traefik_static_config()
+
+        self.static_config["file"] = {"filename": "rules.toml", "watch": True}
+        toml_string = toml.dumps(self.static_config)
+
+        try:
+            with open(self.toml_static_config_file, "w") as f:
+                f.write(toml_string)
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: This can be `toml.dump(self.static_config, f)` instead of calling `toml.dumps` above and `f.write` here

File: jupyterhub_traefik_proxy/toml.py [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+            open(self.toml_dynamic_config_file, "a").close()
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: Add a comment explaining that this is ensuring that the dynamic config file exists.

File: jupyterhub_traefik_proxy/toml.py [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+            frontend_rule_path,
+        ) = traefik_utils.generate_route_keys(self, target, routespec, separator=".")
+
+        data = json.dumps(data)
+        rule = traefik_utils.generate_rule(routespec)
+
+        await self.mutex.acquire()
+        try:
+            self.routes_cache[routespec] = {
+                "backend": [
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: Instead of toml-as-text, let's use a dict here, and `toml.dump` to persist it. We may even want to use `toml.load` to read it in from disk to start.

File: jupyterhub_traefik_proxy/toml.py [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+        routespecs and each value is a dict of the form::
+
+          {
+            'routespec': the route specification ([host]/path/)
+            'target': the target host URL (proto://host) for this route
+            'data': the attached data dict for this route (as specified in add_route)
+          }
+        """
+        all_routes = {}
+        await self.mutex.acquire()
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: This can be easier to keep track of with an [async context mananger](https://docs.python.org/3/glossary.html#term-asynchronous-context-manager), which the mutex is:

```python
async with self.mutex:

File: jupyterhub_traefik_proxy/toml.py [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+
+                'routespec': The normalized route specification passed in to add_route
+                    ([host]/path/)
+                'target': The target host for this route (proto://host)
+                'data': The arbitrary data dict that was passed in by JupyterHub when adding this
+                        route.
+
+            None: if there are no routes matching the given routespec
+        """
+        await self.mutex.acquire()
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: same here

File: tests/test_proxy.py [hunk 18]
```diff
@@ -5,49 +5,116 @@
...
     traefik_port = urlparse(proxy.public_url).port
+    traefik_ip = urlparse(proxy.public_url).hostname
+    backend_port = urlparse(target).port
+    backend_ip = urlparse(target).hostname
 
-    # Check if traefik process is reacheable
+    # Wait until traefik and the backend are ready
     await exponential_backoff(
-        utils.check_host_up, "Traefik not reacheable", ip="localhost", port=traefik_port
+        utils.check_services_ready,
```
Comment (minrk) [inline tests/test_proxy.py]: Since we are parsing a url and then reconstructing it in check_services_ready, maybe check_services_ready should take a list of urls rather than separate ips and ports?

File: jupyterhub_traefik_proxy/toml.py (hunk 12 above)
Comment (GeorgianaElena) [reply jupyterhub_traefik_proxy/toml.py]: I agree that it doesn't look pretty as it is, but the `routes_cache` dict is structured so that we can easily delete and get routes from it. `toml.dump` won't work on the dict as it is. We could modify it to make sense for traefik and `toml.dump`, but the deletion and retrieving of routes would complicate a lot in my opinion. 
I could get rid of the tabs in `add_route` since they don't matter (I've noticed toml.dump doesn't print them) and make the `routes_cache` prettier.

File: jupyterhub_traefik_proxy/traefik_utils.py [hunk 19]
```diff
@@ -1,107 +1,139 @@
...
-    pass  # Not implemented
+def generate_frontend_backend_entry(proxy, frontend_alias):
+    return proxy.etcd_traefik_prefix + "frontends/" + frontend_alias + "/backend"
 
 
-def create_backend_alias_from_url(url):
-    target = urlparse(url)
-    return "jupyterhub_backend_" + target.netloc
+def generate_frontend_rule_entry(proxy, frontend_alias, separator="/"):
+    frontend_rule_entry = (
```
Comment (minrk) [inline jupyterhub_traefik_proxy/traefik_utils.py]: This can be written as:

```python
frontend_rule_entry = separator.join(["frontends", frontend_alias, "routes", "test"])
```

to avoid writing separator over and over again.

File: jupyterhub_traefik_proxy/etcd.py [hunk 20]
```diff
@@ -261,7 +202,7 @@ def _stop_traefik(self):
         try:
             # Check if traefik was launched
             pid = self.traefik_process.pid
```
Comment (minrk) [inline jupyterhub_traefik_proxy/etcd.py]: We shouldn't be ckecing for traefik_process.pid if `self.should_start`. Most of the time, the traefik process will actually not be a subprocess managed by `self.start()`, it will already be running before we launch here.

File: jupyterhub_traefik_proxy/toml.py (hunk 12 above)
Comment (minrk) [reply jupyterhub_traefik_proxy/toml.py]: We need to be able to load state from this file, e.g. when jupyterhub restarts while the proxy is still running. This means that all data (target, routespec, data) must be retrievable from the toml. So in general, we need to have the ability to transform from this toml file to internal routes and back. For that reason, I don't think storing toml strings is the right thing to do. Plus, it is vulenerable to toml syntax errors, which writing toml data directly isn't.

Comment (GeorgianaElena) [reply jupyterhub_traefik_proxy/toml.py]: You're right. I didn't think of this case. Do you agree to add the routespec to the frontend/backend aliases in order to be able to find them easier? Will those be unique? Right now the alias is build based on the target.

File: jupyterhub_traefik_proxy/toml.py [hunk 21]
```diff
@@ -64,9 +64,7 @@ def __init__(self, **kwargs):
         except IOError:
             self.log.exception("Couldn't set up traefik's static config.")
         except:
-            self.log.error(
-                "Couldn't set up traefik's static config. Unexpected error:",
-            )
+            self.log.error("Couldn't set up traefik's static config. Unexpected error:")
```
Comment (minrk) [inline jupyterhub_traefik_proxy/toml.py]: add a `raise` here because we don't want to continue if we can't create the configuration.

File: jupyterhub_traefik_proxy/toml.py (hunk 12 above)
Comment (minrk) [reply jupyterhub_traefik_proxy/toml.py]: In fact, it's probably a good idea to isolate and test the roundtrip from the `get_routes` structure to/from toml to make sure it's stable.

**Comment by GeorgianaElena** [comment]
@minrk, thanks for your feedback. I've pushed some more changes.

**Comment by GeorgianaElena** [comment]
I've added the test. Thanks @minrk ! :smiley_cat:

**Comment by minrk** [comment]
Excellent!
